### ISSUE TITLE
Possible memory leak (again) in generic models after trying V2

### ISSUE THREAD

**Issue by MarkusSintonen** [pydantic/pydantic#6672]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

After taking Pydantic V2 into use some of our tests started to fail that try to ensure there are no leaks related to Pydantic generic models. Below is an example test that succeeds in Pydantic V1 but fails in V2 which might indicate memory leak issues as the models are not necessarily collected properly. Test checks the GC behaviour of a partially parametrized generic model.

(In V1 there was a fix for a similar issue but seems the issue might have popped up again: https://github.com/pydantic/pydantic/pull/5052)

Is V2 safe to take into use because of a possible regression here in garbage collection of models?

Thank you!

### Example Code

```Python
import gc
from typing import Generic, TypeVar

try:
    # v2
    from pydantic._internal._generics import _GENERIC_TYPES_CACHE as cache
    from pydantic import BaseModel as PydanticModel
except ImportError:
    # v1
    from pydantic.generics import GenericModel as PydanticModel, _generic_types_cache as cache


T = TypeVar("T")
C = TypeVar("C")


def test_collects():
    assert len(cache) == 0

    def fn():
        class Inner(PydanticModel, Generic[T, C]):
            a: T
            b: C

        class Outer(PydanticModel, Generic[C]):
            c: Inner[int, C]

        klass = Outer[str]
        assert len(cache) > 0
        assert klass in cache.values()

    fn()

    gc.collect(0)
    gc.collect(1)
    gc.collect(2)

    # Fails in V2 but succeeds in V1!:
    # AssertionError: [<class 'tests.test.Inner[int, ~C]'>, <class 'tests.test.Inner[int, ~C]'>]
    # assert 2 == 0
    assert len(cache) == 0, [*cache.values()]
```

### Python, Pydantic & OS Version

```Text
For V1 (doesnt fail):

             pydantic version: 1.10.11
            pydantic compiled: True
                 install path: /Users/markussintonen/Library/Caches/pypoetry/virtualenvs/leak-test-VvUvdXxD-py3.11/lib/python3.11/site-packages/pydantic
               python version: 3.11.1 (main, Dec 23 2022, 09:40:27) [Clang 14.0.0 (clang-1400.0.29.202)]
                     platform: macOS-13.4.1-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']


For V2 (fails):
             pydantic version: 2.0.2
        pydantic-core version: 2.1.2 release build profile
                 install path: /Users/markussintonen/Library/Caches/pypoetry/virtualenvs/leak-test-VvUvdXxD-py3.11/lib/python3.11/site-packages/pydantic
               python version: 3.11.1 (main, Dec 23 2022, 09:40:27) [Clang 14.0.0 (clang-1400.0.29.202)]
                     platform: macOS-13.4.1-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

Selected Assignee: @davidhewitt

**Comment by MarkusSintonen** [comment]
Here is root-chain object graph showing the possible source of the leak.
```
objgraph.show_chain(
  objgraph.find_backref_chain([*cache.values()][-1], objgraph.is_proper_module),
  filename='chain.png'
)
```

It seems there might be a bug on what the `WeakValueDictionary` gets inside its keys-`tuple`. It seems the weak value dict key has a strong reference to models. 
![chain](https://github.com/pydantic/pydantic/assets/12939780/b1a92f85-ec48-4b0f-874f-10790d98555f)

**Comment by hramezani** [comment]
Thanks @MarkusSintonen for reporting this 🙏 

Could you please test it again with the newly released [Pydantic 2.0.3](https://pypi.org/project/pydantic/2.0.3/)?

**Comment by dmontagu** [comment]
@MarkusSintonen it won't surprise me if there is some memory leak coming from `__pydantic_parent_namespace__` which I believe was added to try to make auto-rebuilds work better. Worst case I think we could remove that and go back to requiring more "manual" calls to `model_rebuild`.

Thank you for the reproducible example, we should probably add it at least as an xfailing test.

**Comment by dmontagu** [comment]
I'm thinking maybe if we make the parent namespace a weakdict or in some other way make it more dynamically accessed/accessible it might fix this... I'll play with it a bit

**Comment by dmontagu** [comment]
@MarkusSintonen I'd be interested in your feedback on #6681

**Comment by dmontagu** [comment]
@MarkusSintonen Also, I just want to add that that diagram made it 1000x easier for me to get to the bottom of this, thank you for sharing that.

**Comment by MarkusSintonen** [comment]
> @MarkusSintonen I'd be interested in your feedback on #6681

Thank you for the quick fix! Threw there couple question/comments.

**Comment by MarkusSintonen** [comment]
> Thanks @MarkusSintonen for reporting this 🙏
> 
> Could you please test it again with the newly released [Pydantic 2.0.3](https://pypi.org/project/pydantic/2.0.3/)?

Its still happening in 2.0.3:
```
poetry run python -c "import pydantic.version; print(pydantic.version.version_info())"
        pydantic version: 2.0.3
        pydantic-core version: 2.3.0 release build profile
                 install path: /Users/markussintonen/Library/Caches/pypoetry/virtualenvs/leak-test-VvUvdXxD-py3.11/lib/python3.11/site-packages/pydantic
               python version: 3.11.1 (main, Dec 23 2022, 09:40:27) [Clang 14.0.0 (clang-1400.0.29.202)]
                     platform: macOS-13.4.1-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```
